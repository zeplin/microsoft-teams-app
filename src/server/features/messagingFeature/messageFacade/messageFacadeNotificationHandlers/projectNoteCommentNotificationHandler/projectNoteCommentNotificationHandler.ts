import { NotificationHandler } from "../NotificationHandler";
import { MessageCard, commonTeamsCard } from "../teamsCardTemplates";
import { ProjectNoteCommentEvent, WebhookEvent } from "../../../../../adapters/zeplin/types";
import { ZEPLIN_MAC_APP_URL_SCHEME, ZEPLIN_WEB_APP_BASE_URL } from "../../../../../config";
import { getMacAppRedirectURL } from "../getMacAppRedirectURL";

class ProjectNoteCommentNotificationHandler extends NotificationHandler {
    delay: 0;
    private getText(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                context: {
                    project: {
                        name: projectName
                    },
                    screen: {
                        name: screenName
                    },
                    note: {
                        order
                    }
                },
                actor: {
                    user: {
                        username
                    }
                }
            }
        } = event;
        return `**${username}** replied to note _#${order}_ on _${screenName}_ screen in _${projectName}_. üèÉ‚Äç‚ôÇ`;
    }

    private getSectionText(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                resource: {
                    data: {
                        content: commentContent
                    }
                }
            }
        } = event;
        return commentContent;
    }

    private getWebappURL(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                context: {
                    screen: {
                        id: screenId
                    },
                    project: {
                        id: projectId
                    },
                    note: {
                        id: noteId
                    }
                },
                resource: {
                    id: commentId
                }
            }
        } = event;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/screen/${screenId}`;
        webappURL.searchParams.set("did", noteId);
        webappURL.searchParams.set("cmid", commentId);
        return webappURL.toString();
    }

    private getMacAppURL(event: ProjectNoteCommentEvent): string {
        const {
            payload: {
                context: {
                    project: {
                        id: projectId
                    },
                    screen: {
                        id: screenId
                    },
                    note: {
                        id: noteId
                    }
                },
                resource: {
                    id: commentId
                }
            }
        } = event;
        return getMacAppRedirectURL(`${ZEPLIN_MAC_APP_URL_SCHEME}://dot?pid=${projectId}&sid=${screenId}&did=${noteId}&cmids=${commentId}`);
    }

    shouldHandleEvent(event: WebhookEvent): boolean {
        return event.payload.action === "created";
    }

    getGroupingKey(event: WebhookEvent): string {
        return event.deliveryId;
    }

    getTeamsMessage(events: ProjectNoteCommentEvent[]): MessageCard {
        const [event] = events;
        return commonTeamsCard({
            text: this.getText(event),
            section: {
                text: this.getSectionText(event)
            },
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(event)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(event)
            }]
        });
    }
}

export const projectNoteCommentNotificationHandler = new ProjectNoteCommentNotificationHandler();
