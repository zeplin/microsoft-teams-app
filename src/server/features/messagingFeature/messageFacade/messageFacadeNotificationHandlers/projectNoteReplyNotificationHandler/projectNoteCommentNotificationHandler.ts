import { NotificationHandler } from "../NotificationHandler";
import { AdaptiveCard, commonTeamsCard } from "../teamsCardTemplates";
import {
    WebhookEvent,
    EventType,
    EventPayload,
    ProjectContext,
    ScreenContext,
    NoteContext
} from "../../../messagingTypes";
import { ScreenNoteCommentResource } from "../resources/screenNoteCommentResource";
import { ZEPLIN_MAC_APP_URL_SCHEME, ZEPLIN_WEB_APP_BASE_URL } from "../../../../../config";

type ProjectNoteCommentEventDescriptor = {
    type: EventType.PROJECT_NOTE_COMMENT;
    action: "created";
}

class ProjectNoteCommentNotificationHandler extends NotificationHandler {
    delay: 0;

    private getWebappURL(event: WebhookEvent<ProjectNoteCommentEventPayload>): string {
        const {
            payload: {
                context: {
                    screen: {
                        id: screenId
                    },
                    project: {
                        id: projectId
                    },
                    note: {
                        id: noteId
                    }
                },
                resource: {
                    id: commentId
                }
            }
        } = event;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/screen/${screenId}`;
        webappURL.searchParams.set("did", noteId);
        webappURL.searchParams.set("cmid", commentId);
        return webappURL.toString();
    }

    getGroupingKey(event: WebhookEvent): string {
        return event.deliveryId;
    }

    getTeamsMessage(events: WebhookEvent<ProjectNoteCommentEventPayload>[]): AdaptiveCard {
        if (events.length > 1) {
            // TODO: Log that there was somehow grouping done for this notification
        }

        const [event] = events;
        const {
            payload: {
                context: {
                    project: {
                        id: projectId,
                        name: projectName
                    },
                    screen: {
                        id: screenId,
                        name: screenName
                    },
                    note: {
                        id: noteId,
                        order
                    }
                },
                resource: {
                    id: commentId,
                    data: {
                        content: commentContent
                    }
                },
                actor: {
                    user: {
                        username
                    }
                }
            }
        } = event;
        const cardText = `**${username}** replied to note _#${order}_ on _${screenName}_ screen in _${projectName}_. üèÉ‚Äç‚ôÇ`;
        return commonTeamsCard({
            text: cardText,
            sectionText: commentContent,
            links: [{
                title: "Open in App",
                url: `${ZEPLIN_MAC_APP_URL_SCHEME}dot?pid=${projectId}&sid=${screenId}&did=${noteId}&cmids=${commentId}`
            }, {
                title: "Open in Web",
                url: this.getWebappURL(event)
            }]
        });
    }
}

export type ProjectNoteCommentEventPayload = EventPayload<
    ProjectNoteCommentEventDescriptor,
    ProjectContext & ScreenContext & NoteContext,
    ScreenNoteCommentResource
>;
export const projectNoteCommentNotificationHandler = new ProjectNoteCommentNotificationHandler();