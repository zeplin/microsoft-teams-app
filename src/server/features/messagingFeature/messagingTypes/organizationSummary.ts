export interface OrganizationSummary {
    id: string;
    name: string;
    logo?: string;
}