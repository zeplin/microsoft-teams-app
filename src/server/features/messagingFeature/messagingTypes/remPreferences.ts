export type RemPreferences = {
    status: "enabled" ;
    root_font_size: number;
    use_for_font_sizes: boolean;
    use_for_measurements: boolean;
} | { status: "disabled" | "linked" }