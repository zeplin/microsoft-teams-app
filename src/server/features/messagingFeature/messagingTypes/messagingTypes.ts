import { OrganizationSummary } from "./organizationSummary";
import { RemPreferences } from "./remPreferences";

type EventDescriptor = {
    type: EventType;
    action: string;
}

type Resource = {
    id: string;
    type: ResourceType;
    data?: object;
}

export enum EventType {
    PROJECT_COLOR = "project.color",
    STYLEGUIDE_COLOR = "styleguide.color",
    PROJECT_NOTE = "project.note"
}

export interface ProjectContext {
    project: {
        id: string;
        created: number;
        name: string;
        description?: string;
        thumbnail?: string;
        platform: "android" | "ios" | "web" | "macos";
        status: "active" | "archived" | "deleted";
        organization?: OrganizationSummary;
        scene_url?: string;
        updated?: number;
        rem_preferences?: RemPreferences;
        number_of_screens: number;
        number_of_components: number;
        number_of_text_styles: number;
        number_of_colors: number;
        number_of_spacing_tokens: number;
        number_of_members: number;
        linked_styleguide?: {
            id: string;
        };
    };
}

export interface ScreenContext {
    screen: {
        id: string;
        created: number;
        updated?: number;
        tags: string[];
        name: string;
        image: {
            width: number;
            height: number;
            original_url?: string;
        };
        section?: {
            id: string;
        };
        description?: string;
        number_of_versions: number;
        number_of_notes: number;
    };
}

export interface StyleguideContext {
    styleguide: {
        id: string;
        created: number;
        name: string;
        description?: string;
        thumbnail?: string;
        platform: "base" | "web" | "ios" | "android" | "macos";
        status: "active" | "archived" | "deleted";
        organization?: OrganizationSummary;
        updated?: number;
        rem_preferences?: RemPreferences;
        number_of_components: number;
        number_of_text_styles: number;
        number_of_colors: number;
        number_of_spacing_tokens: number;
        number_of_members: number;
        parent?: {
            id: string;
        };
    };
}

export enum ResourceType {
    PROJECT = "Project",
    STYLEGUIDE = "Styleguide",
    COLOR = "Color",
    TEXT_STYLE = "TextStyle",
    COMPONENT = "Component",
    SPACING_TOKEN = "SpacingToken",
    PROJECT_MEMBER = "ProjectMember",
    STYLEGUIDE_MEMBER = "StyleguideMember",
    ORGANIZATION_SUMMARY = "OrganizationSummary",
    SCREEN = "Screen",
    SCREEN_VERSION = "ScreenVersion",
    ORGANIZATION_MEMBER = "OrganizationMember",
    SCREEN_NOTE = "ScreenNote",
    SCREEN_NOTE_COMMENT = "ScreenNoteComment"
}

export type User = {
    id: string;
    email: string;
    username: string;
    emotar?: string;
    avatar?: string;
}

export interface EventPayload<
    E extends EventDescriptor,
    C extends ProjectContext | StyleguideContext,
    R extends Resource,
> {
    event: E["type"];
    action: E["action"];
    timestamp: number;
    context: C;
    resource: R;
    actor: {
        user: User;
    };
}

export type CommonEventPayload = EventPayload<EventDescriptor, ProjectContext | StyleguideContext, Resource>;

export type WebhookEvent<T extends CommonEventPayload = CommonEventPayload> = {
    webhookId: string;
    deliveryId: string;
    payload: T;
};

export type MessageJobData = {
    id: string;
    groupingKey: string;
}