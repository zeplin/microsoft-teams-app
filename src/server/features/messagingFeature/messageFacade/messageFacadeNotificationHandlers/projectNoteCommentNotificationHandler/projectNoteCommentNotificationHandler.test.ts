import { WebhookEvent } from "../../../messagingTypes";
import { ProjectNoteCommentEventPayload, projectNoteCommentNotificationHandler } from "./projectNoteCommentNotificationHandler";

type GetDummyEventParams = {
    projectId?: string;
    projectName?: string;
    screenId?: string;
    screenName?: string;
    username?: string;
    noteId?: string;
    noteOrder?: string;
    commentId?: string;
    commentContent?: string;
}

function getDummyEvent({
    projectId = "projectId",
    projectName = "Project MS Teams Integration",
    screenId = "screenId",
    screenName = "Manage Zeplin Connector",
    username = "sertac",
    noteId = "noteId",
    noteOrder = "1",
    commentId = "commentId",
    commentContent = "cok guzel keyfimiz var"
}: GetDummyEventParams = {}): WebhookEvent<ProjectNoteCommentEventPayload> {
    return {
        payload: {
            action: "created",
            context: {
                screen: {
                    id: screenId,
                    name: screenName
                },
                project: {
                    id: projectId,
                    name: projectName
                },
                note: {
                    id: noteId,
                    order: noteOrder
                }
            },
            actor: {
                user: {
                    username
                }
            },
            resource: {
                id: commentId,
                data: {
                    id: commentId,
                    content: commentContent
                }
            }
        }
    } as WebhookEvent<ProjectNoteCommentEventPayload>;
}

describe("projectNoteNotificationHandler", () => {
    describe("getTeamsMessage", () => {
        it("should match snapshot", () => {
            expect(projectNoteCommentNotificationHandler.getTeamsMessage([getDummyEvent()])).toMatchSnapshot();
        });
    });
});