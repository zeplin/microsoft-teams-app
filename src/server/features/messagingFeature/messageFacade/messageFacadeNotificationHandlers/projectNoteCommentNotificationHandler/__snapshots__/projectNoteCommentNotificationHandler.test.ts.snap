// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectNoteNotificationHandler getTeamsMessage should match snapshot 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**sertac** replied to note _#1_ on _Manage Zeplin Connector_ screen in _Project MS Teams Integration_. üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "cok guzel keyfimiz var",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "zpl://dot?pid=projectId&sid=screenId&did=noteId&cmids=commentId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/screen/screenId?did=noteId&cmid=commentId"
        }
      ]
    }
  ]
}
`;
